salvo/src/lib/benchmark/binary_benchmark.py

```diff
@@ -0,0 +1,174 @@
+"""
+This module contains the methods to perform a Binary benchmark using
+containers for the scripts, nighthawk binaries, and envoy
+
+https://github.com/envoyproxy/nighthawk/blob/master/benchmarks/README.md
+"""
+
+import subprocess
+import logging
+import os
+
+import api.control_pb2 as proto_control
+import api.source_pb2 as proto_source
+
+from src.lib.benchmark import base_benchmark
+from src.lib.builder import (envoy_builder, nighthawk_builder)
+from src.lib import (cmd_exec, source_manager)
+
+log = logging.getLogger(__name__)
+
+
+class BinaryBenchmarkError(Exception):
+  """Error raised when running a binary benchmark in cases
+     where we cannot make progress due to abnormal conditions.
+  """
+
+class Benchmark(base_benchmark.BaseBenchmark):
+  """This benchmark class is the binary benchmark. We use a path to an Envoy
+     binary to execute the Nighthawk benchmarks using that specific build.
+  """
+
+  def __init__(
+    self, job_control: proto_control.JobControl, benchmark_name: str) -> None:
+    """ Initialize the benchmark class."""
+
+    super(Benchmark, self).__init__(job_control, benchmark_name)
+    self._benchmark_dir = None
+    self.envoy_binary_path = job_control.environment.variables['ENVOY_PATH']
+    self.envoy_builder = None
+    self.nighthawk_builder = None
+    self.source_manager = source_manager.SourceManager(job_control)
+    self.use_fallback = False
+
+  def _validate(self) -> None:
+    """Validate that all data required for running a benchmark exists.
+
+    Verify that a source has been specified with which to build Nighthawk,
+    warns the user if no valid Envoy binary is specified.
+
+    Returns:
+        None
+    """
+
+    source = self.get_source()
+    if not source:
+      raise BinaryBenchmarkError("No source configuration specified")
+
+    can_build_envoy = False
+    can_build_nighthawk = False
+
+    for source_def in source:
+      if source_def.identity == source_def.SRCID_UNSPECIFIED:
+        raise BinaryBenchmarkError("No source identity specified")
+
+      if source_def.identity == source_def.SRCID_ENVOY and \
+        (source_def.source_path or source_def.source_url):
+        can_build_envoy = True
+
+      if source_def.identity == source_def.SRCID_NIGHTHAWK and \
+        (source_def.source_path or source_def.source_url):
+        can_build_nighthawk = True
+
+    if not can_build_nighthawk:
+      raise BinaryBenchmarkError("No source specified to build Nighthawk")
+
+    if not (can_build_envoy or self.envoy_binary_path):
+      self.use_fallback = True
```

Comment:
Now that I'm taking another look at the fallback functionality, I'm starting to realize that removing it altogether is the best course of action. Technically, nighthawk *can* execute without an envoy binary, using its built-in test server, but that doesn't give any tangible benefit for the kinds of workloads that Salvo is expected to automate.

Additionally, the frontend (i.e. `run_benchmark.py`) will already fail-fast when no Envoy source is provided.

This is essentially a vestigial feature from the initial development of the Binary Benchmark workflow, and your comments highlight to me that it is more appropriate to fail the workflow in this case.

Reply:
Re: Now that I'm taking another look at the fallback functionality, I'm starting to realize that removing it altogether is the best course of action. Technically, nighthawk can execute without an envoy binary, using its built-in test server, but that doesn't give any tangible benefit for the kinds of workloads that Salvo is expected to automate.

Drive by comment: it might serve a purpose in baselining?

Reply:
> it might serve a purpose in baselining?

That is a possibility, though any particular version of Envoy could also be used as a baseline.
If we want to keep the fallback logic, I'd probably have it be controlled by an environment variable in the job control.

If this sounds like a useful feature, I can add it - but I kind of doubt that people would use it at the moment.

Reply:
So, to illustrate, with baseline I meant `nh client -> nh test server` and then one can compare against `nh client -> envoy -> nh test server`. To see how introducing an Envoy proxy affects measurements.

If it's extra work, that's not my intent, it's just that the comment stood out, and it seemed like something would be axed out that I thought could serve a purpose.

Reply:
I see. That does make sense, and I can see now how that would add value.
Unfortunately, this feature as it exists is not complete on an end-to-end basis, and would require additional development and testing on the frontend/job control side in order to get it working.

I think this is a good callout, but I think the best course of action is to remove it for now.

Reply:
I would rely on you for judging that, but perhaps it is good to file an issue for it so it can be tracked.

The benchmark scripts support the flow, for example:
https://github.com/envoyproxy/nighthawk/blob/main/benchmarks/test/test_discovery.py#L101-L105
(which produces https://27055-180498819-gh.circle-artifacts.com/0/2021-03-29-18-55-03/http_h1_small_request_small_reply_direct_IpVersion.IPV4-nighthawk_test_integration_configurations_nighthawk_http_origin.yaml/simple_plot.html in CI)

Reply:
Sorry, thinking about this more, one afterthought.
The above discussion represents a case where there's 0 Envoy instances.
If it impacts things here, >1 Envoy instances (from the same version/binary) might be good to consider too in the context of this: a chain of Envoy's is a realistic scenario, for example gateway->sidecar, and therefore could be interesting to test at some point.

Reply:
Agreed that dropping this functionality from this PR is a good idea. @oschaaf thank you for raising a good point about the need for having a baseline check for the Salvo/Nighthawk combo itself. Otto, if you have the cycles, I would ask for your help with filing of the issue describing this functionality (both 0 Envoys and >1 Envoys). While the functionality isn't on the critical path that we are currently developing, it is something worth exploring and tracking in the future.